src/progress.rs

```diff
@@ -278,25 +283,42 @@ impl Progress {
     /// Returns false if the given index comes from an out of order message.
     /// Otherwise it decreases the progress next index to min(rejected, last)
     /// and returns true.
-    pub fn maybe_decr_to(&mut self, rejected: u64, last: u64) -> bool {
+    pub fn maybe_decr_to(&mut self, rejected: u64, last: u64, request_snapshot: u64) -> bool {
         if self.state == ProgressState::Replicate {
             // the rejection must be stale if the progress has matched and "rejected"
             // is smaller than "match".
-            if rejected <= self.matched {
+            // Or rejected equals to matched and request_snapshot is the INVALID_INDEX.
+            if rejected < self.matched
+                || (rejected == self.matched && request_snapshot == INVALID_INDEX)
+            {
                 return false;
             }
             self.next_idx = self.matched + 1;
+            if request_snapshot == INVALID_INDEX {
+                self.next_idx = self.matched + 1;
```

Comment:
```rust
self.next_idx = self.matched + 1;
if request_snapshot == INVALID_INDEX { 
    self.next_idx = self.matched + 1; 
```

Are you sure? I think there's a better way of writing this.

Reply:
I just applied the diff and fixed it up. This is from https://github.com/pingcap/raft-rs/pull/243/files#diff-2731f87d144f80edb31dac3d81fd9107R296

Reply:
This line was deleted https://github.com/pingcap/raft-rs/pull/243/files#diff-2731f87d144f80edb31dac3d81fd9107L288

So in your case you should probably delete line 288 too

Reply:
Oh right, d'oh. Thanks